### PR TITLE
PACT-809 Create unit test suite for Login view

### PR DESCRIPTION
**Pull request by userjd**

### CODE REVIEW

**Comment by tna-erasmos** [comment]
@userjd , could we reformat this?

Just a reminder that we have `runBeforePushing.sh`.

File: test/views/LoginViewSpec.scala [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+import support.ExpectedValues.ExpectedSummaryErrorMessage
+import uk.gov.hmrc.govukfrontend.views.html.components.{ GovukButton, GovukErrorSummary, GovukFieldset }
+import uk.gov.nationalarchives.omega.editorial.forms.CredentialsFormProvider
+import uk.gov.nationalarchives.omega.editorial.models.Credentials
+import uk.gov.nationalarchives.omega.editorial.views.html.login
+
+class LoginViewSpec extends BaseViewSpec {
+  private val errorSummaryTitle = "There is a problem"
+  val defaultLang = play.api.i18n.Lang.defaultLang.code
+  private val messages: Map[String, Map[String, String]] = Map(
```
Comment (tna-erasmos) [inline test/views/LoginViewSpec.scala:39]: Why are we asserting message keys in some places and the resolved messages in others? 

I'd suggest just asserting the keys, as the resolution is handled in the integration tests.

**Review by tna-erasmos — COMMENTED** [review]
Minor suggestions.

File: test/views/LoginViewSpec.scala (hunk 1 above)
Comment (userjd) [reply test/views/LoginViewSpec.scala]: Removed

File: test/views/LoginViewSpec.scala [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+    defaultLang -> Map(
+      "error.summary.title" -> errorSummaryTitle
+    )
+  )
+  implicit val messagesApi: MessagesApi = stubMessagesApi(messages)
+
+  "Login Html" should {
+    "render the given title and heading" in {
+      val credentialsForm: Form[Credentials] = CredentialsFormProvider()(FakeRequest(), messagesApi)
+      val loginHtml =
```
[the remark this answers is not in the document]
Comment (userjd) [reply test/views/LoginViewSpec.scala:45]: Done

File: test/views/LoginViewSpec.scala [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+        .withError(FormError("username", "Enter a username"))
+        .withError(FormError("password", "Enter a password"))
+      val loginHtml =
+        getLoginInstance(userForm)(Helpers.stubMessages(messagesApi), CSRFTokenHelper.addCSRFToken(FakeRequest()))
+      val document = asDocument(loginHtml)
+      document must haveSummaryErrorMessages(
+        ExpectedSummaryErrorMessage("Enter a username", "username"),
+        ExpectedSummaryErrorMessage("Enter a password", "password")
+      )
+      document must haveErrorMessageForUsername("login.hidden.error Enter a username")
```
[the remark this answers is not in the document]
Comment (userjd) [reply test/views/LoginViewSpec.scala]: Done

File: test/views/LoginViewSpec.scala [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+      val document = asDocument(loginHtml)
+      document must haveSummaryErrorTitle(errorSummaryTitle)
+      document must haveSummaryErrorMessages(
+        ExpectedSummaryErrorMessage("Username and/or password is incorrect.", "username")
+      )
+    }
+
+    "render an error when no username given" in {
+      val credentialsForm: Form[Credentials] = CredentialsFormProvider()(FakeRequest(), messagesApi)
+        .fill(Credentials.apply("", ""))
```
[the remark this answers is not in the document]
Comment (userjd) [reply test/views/LoginViewSpec.scala]: Done

File: test/views/LoginViewSpec.scala [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+        )
+
+      val document = asDocument(loginHtml)
+      document must haveSummaryErrorTitle(errorSummaryTitle)
+      document must haveSummaryErrorMessages(ExpectedSummaryErrorMessage("Enter a password", "password"))
+      document must haveErrorMessageForPassword("login.hidden.error Enter a password")
+    }
+  }
+
+  private def getLoginInstance: login =
```
[the remark this answers is not in the document]
Comment (userjd) [reply test/views/LoginViewSpec.scala]: Done

File: test/views/LoginViewSpec.scala [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+        getLoginInstance(credentialsForm)(
+          Helpers.stubMessages(messagesApi),
+          CSRFTokenHelper.addCSRFToken(FakeRequest())
+        )
+
+      val document = asDocument(loginHtml)
+      document must haveSummaryErrorTitle(errorSummaryTitle)
+      document must haveSummaryErrorMessages(ExpectedSummaryErrorMessage("Enter a username", "username"))
+      document must haveErrorMessageForUsername("login.hidden.error Enter a username")
+    }
```
[the remark this answers is not in the document]
Comment (userjd) [reply test/views/LoginViewSpec.scala:113]: As the login form values are reset it wouldn't be possible to assert against the input values.

File: test/views/LoginViewSpec.scala (hunk 1 above)
Comment (tna-erasmos) [reply test/views/LoginViewSpec.scala]: Confirmed.

File: test/views/LoginViewSpec.scala (hunk 3 above)
[the remark this answers is not in the document]
Comment (tna-erasmos) [reply test/views/LoginViewSpec.scala]: Confirmed.

File: test/views/LoginViewSpec.scala (hunk 6 above)
[the remark this answers is not in the document]
Comment (tna-erasmos) [reply test/views/LoginViewSpec.scala:113]: That's not always the case; for instance in "render an error when no password given":

```
document must haveUsername("11")
document must havePassword("")
```

And where both are blanked out, we can demonstrate that, can't we? In "render an error when no username given" 👍 

```
document must haveUsername("")
document must havePassword("")
```

File: test/views/LoginViewSpec.scala (hunk 5 above)
[the remark this answers is not in the document]
Comment (tna-erasmos) [reply test/views/LoginViewSpec.scala]: Confirmed.

File: test/views/LoginViewSpec.scala (hunk 2 above)
[the remark this answers is not in the document]
Comment (tna-erasmos) [reply test/views/LoginViewSpec.scala]: Confirmed.

File: test/views/LoginViewSpec.scala (hunk 6 above)
[the remark this answers is not in the document]
Comment (tna-erasmos) [reply test/views/LoginViewSpec.scala:101]: Close, but not quite: we can assert both username and password in all of the scenarios.

[the remark this answers is not in the document]
Comment (userjd) [reply test/views/LoginViewSpec.scala:106]: Done

[the remark this answers is not in the document]
Comment (tna-erasmos) [reply test/views/LoginViewSpec.scala:106]: Confirmed.

**Pull request merged**